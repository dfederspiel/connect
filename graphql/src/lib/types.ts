export interface IDataContext<T> {
  getAll(): Promise<T[]>;
  get(id: string): Promise<T | null>;
  post(item: T): Promise<T>;
  //delete(id: string): Promise<T>;
  //put(item: T): Promise<T>;
}

export interface IPubSub {
  publish(triggerName: string, payload: any): Promise<void>;
  subscribe(trigger: string, onMessage: any, options?: unknown): Promise<number>;
  asyncIterator<T>(triggers: string | string[]): AsyncIterator<T, any, undefined>;
}
