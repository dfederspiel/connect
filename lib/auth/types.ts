export interface IDataContext<T> {
  getAll(): Promise<T[]>;
  getById(id: number): Promise<T | null>;
  post(item: T): Promise<T>;
  //delete(id: string): Promise<T>;
  //put(item: T): Promise<T>;
}

export interface IPubSub {
  publish(triggerName: string, payload: any): Promise<void>;
  asyncIterator<T>(triggers: string | string[]): AsyncIterator<T, any, undefined>;
}
