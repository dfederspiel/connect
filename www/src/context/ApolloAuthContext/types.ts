import { ApolloClient, NormalizedCacheObject } from '@apollo/client';

export type ApolloAuthContext = {
  client: ApolloClient<NormalizedCacheObject>;
};
