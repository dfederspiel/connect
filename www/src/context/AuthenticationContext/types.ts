import { AuthStatus } from './useProvideAuth';

type Policy = {
  policyId: number;
  url: string;
};

type UserProfile = {
  name: string;
  email: string;
};

type AuthContext = {
  status: AuthStatus;
  user: string | undefined;
  login(): void;
  logout(): void;
  token(forceRefresh: boolean): Promise<string | null>;
};

export type { AuthContext, Policy, UserProfile };
