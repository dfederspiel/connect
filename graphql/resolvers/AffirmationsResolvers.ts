import { User } from '@prisma/client';
import { gql, IResolvers, withFilter } from 'apollo-server-express';
import { IPubSub } from '../lib/types';

const AFFIRMATION_GIVEN = 'AFFIRMATION_GIVEN';

export const AffirmationsTypeDefs = gql`
  type Affirmation {
    from: ID!
    to: ID!
  }

  extend type Mutation {
    sendAffirmation(userId: ID!): Affirmation
  }

  extend type Subscription {
    affirmationGiven: Affirmation
  }
`;

export default class AffirmationsResolvers {
  resolvers: IResolvers;
  pubsub: IPubSub;

  constructor(pubsub: IPubSub) {
    this.resolvers = this.initializeResolvers();
    this.pubsub = pubsub;
  }

  initializeResolvers = (): IResolvers => {
    return {
      Mutation: {
        sendAffirmation: async (_: any, { userId }: any, context: { user: User }) => {
          const affirmation = {
            from: context.user.id,
            to: userId,
          };
          this.pubsub.publish(AFFIRMATION_GIVEN, affirmation);
          return affirmation;
        },
      },
      Subscription: {
        affirmationGiven: {
          subscribe: withFilter(
            () => this.pubsub.asyncIterator('AFFIRMATION_GIVEN'),
            (payload, _, context) => {
              return parseInt(payload.to) === context.user.id;
            },
          ),
        },
      },
    };
  };
}
