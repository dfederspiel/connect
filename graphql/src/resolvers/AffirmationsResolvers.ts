import { User } from '@prisma/client';
import { gql, withFilter } from 'apollo-server-express';
import { IPubSub } from '../lib/types';

const AFFIRMATION_GIVEN = 'AFFIRMATION_GIVEN';

export const AffirmationsTypeDefs = gql`
  type Affirmation {
    from: ID!
    to: ID!
  }

  extend type Mutation {
    sendAffirmation(userId: ID!): Affirmation
  }

  extend type Subscription {
    affirmationGiven: Affirmation
  }
`;

interface IAffirmationsMutationResolvers {
  sendAffirmation(
    parent: any,
    args: any,
    context: any,
    other: any,
  ): Promise<{
    from: number;
    to: any;
  }>;
}

interface IAffirmationsSubscriptionResolvers {
  affirmationGiven: {
    subscribe(): any;
  };
}
interface IAffirmationsResolvers {
  Mutation: IAffirmationsMutationResolvers;
  Subscription: IAffirmationsSubscriptionResolvers;
}

export default class AffirmationsResolvers {
  resolvers: IAffirmationsResolvers;
  pubsub: IPubSub;

  constructor(pubsub: IPubSub) {
    this.resolvers = this.initializeResolvers();
    this.pubsub = pubsub;
  }

  initializeResolvers = (): IAffirmationsResolvers => {
    return {
      Mutation: {
        sendAffirmation: async (_: any, { userId }: any, context: { user: User }) => {
          const affirmation = {
            from: context.user.id,
            to: userId,
          };
          this.pubsub.publish(AFFIRMATION_GIVEN, affirmation);
          return affirmation;
        },
      },
      Subscription: {
        affirmationGiven: {
          subscribe: withFilter(
            () => this.pubsub.asyncIterator('AFFIRMATION_GIVEN'),
            (payload, _args, context) => {
              return parseInt(payload.to) === context.user.id;
            },
          ),
        },
      },
    };
  };
}
