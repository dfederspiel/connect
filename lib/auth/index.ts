export { AuthContext } from '../auth/AuthContext';
export { IUserDataContext } from '../auth/UserDataContext';
export { IDataContext, IPubSub } from '../auth/types';
