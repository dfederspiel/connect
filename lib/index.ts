export { AuthContext } from './AuthContext';
export { IUserDataContext } from './UserDataContext';
export { IDataContext, IPubSub } from './types';

export const test = (source: string): string => {
  return `ALIASING WORKING FOR ${source}`;
};
