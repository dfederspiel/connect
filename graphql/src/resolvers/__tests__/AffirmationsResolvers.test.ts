/* eslint-disable prettier/prettier */
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/no-empty-function */
import { User } from '@prisma/client';
import { IPubSub } from '../../lib/types';
import AffirmationsResolvers from '../AffirmationsResolvers';


const mockPubSub = {
  publish: (tag, payload) => { },
  asyncIterator: (tag) => { },
} as IPubSub;
const context = {
  user: {
    id: 1,
    domain: '',
    email: '',
  } as User,
};

describe('the affirmations resolver', () => {
  it('exists', () => {
    const resolver = new AffirmationsResolvers(mockPubSub);
    expect(resolver).toBeDefined();
  });

  it('can send affirmations', async () => {
    const resolver = new AffirmationsResolvers(mockPubSub);
    const result = await resolver.resolvers.Mutation.sendAffirmation(
      null,
      { userId: '1' },
      context,
      null,
    );
    expect(result).toMatchSnapshot();
  });

  xit('can cover subscriptions', async () => {
    const resolver = new AffirmationsResolvers(mockPubSub as unknown as IPubSub);
    const result = await resolver.resolvers.Subscription.affirmationGiven.subscribe();
    await result.return({
      to: '1',
    }, null, {
      user: {
        id: '1'
      }
    });
    expect(result).toBeDefined();
  });
});
