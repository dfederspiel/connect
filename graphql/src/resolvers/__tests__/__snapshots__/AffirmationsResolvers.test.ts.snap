// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the affirmations resolver can send affirmations 1`] = `
Object {
  "from": 1,
  "to": "1",
}
`;
